src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/RateLimiters/IRateLimiter.cs

```diff
@@ -0,0 +1,37 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Data.SqlClientX.RateLimiters
+{
+    /// <summary>
+    /// An interface for rate limiters that execute arbitraty code. Intended to be small and self contained and chained together to achieve more complex behavior.
+    /// </summary>
+    internal abstract class IRateLimiter
```

Comment:
100% the long term idea behind this pattern. For now, I'm biasing towards leaving new functionality as internal. We need to be careful with the public API we expose because we're required to update docs and other things appropriately, so just want to cross that bridge when we get to it.

Reply:
Of note, ResiliencePipelineT in polly understandably requires access to the return type of the pipeline.
https://github.com/App-vNext/Polly/blob/main/src/Polly.Core/ResiliencePipelineT.Async.cs#L5
So to integrate with Polly, we'd need to expose the SqlConnector type which should be an implementation detail of the driver. There may be some solution with an added layer of indirection? But it's not my top focus right now. Most customers are looking for fewer/no rate limits, not the ability to customize their rate limiting.

Reply:
Your point around `public` makes sense, and I agree that integration with a third-party rate limiter is a little distant right now.

I'll wait to see the way the `BlockingPeriodRateLimiter` works before making any firm suggestion, but I agree with not exposing `SqlConnector`. Maybe we could make this work by instantiating a new instance of a `RateLimiterBase`-derived class which calls a third party rate limiter, passing a delegate internal to the instance of the derived class which creates a SqlConnector and assigns it to a field - something vaguely similar to the below, if we can make the generics work (or remove them.)

```csharp
public interface IResiliencyHook
{
    Task ExecuteAsync(Func<ValueTask> onConnect, SqlConnectionStringBuilder connectionString, CancellationToken cancellationToken);

    void Execute(Action onConnect, SqlConnectionStringBuilder connectionString, CancellationToken cancellationToken);
}

internal sealed class ExternalRateLimiter<TResiliencyHook> : RateLimiterBase
    where TResiliencyHook : IResiliencyHook
{
    private SqlConnector _resultantConnector;
    private TResiliencyHook _hook;

    public ExternalRateLimiter (TResiliencyHook hook)
    {
        _hook = hook;
    }

    internal override ValueTask<TResult> Execute<State, TResult>(AsyncFlagFunc<State, ValueTask<TResult>> callback, State state, bool async, CancellationToken cancellationToken = default)
    {
        if (async)
        {
            await _hook.ExecuteAsync(...);
        }
        else
        {
            _hook.Execute(...);
        }

        return _resultantConnector;
    }
}
```

Reply:
The BlockingPeriodRateLimiter implementation will be covered by #2677